### PR TITLE
Remove unnecessary redistribution node

### PR DESCRIPTION
**Pull request by arzuschen**
Previously join force redistribution on top of table scan for distributed table, but it is not necessary when the key match the distribution.
Now the match is checked, only when the data distribution does not agree with the join keys, a redistribution node is added. 

Also, data insertion is modified. Previously, even for distributed table all the rows are added onto the first partition (0), but with distributed table scan all have redistribution node on top, it is still working. Now redistribution may not be present, the original insertion must follow the distributed column.

Additional unittest is added to verify the change.

### CODE REVIEW

**Comment by arzuschen** [comment]
updated to take in consideration for different query dop.
Since the dop for data insertion and dop for query may not be the same, need to force redistribution if the partition number is not the same.

File: qpmodel/DataType.cs [hunk 1]
```diff
@@ -261,6 +261,23 @@ public BaseTableRef(string name, string alias = null, SelectStmt.TableSample tab
         }
 
         public bool IsDistributed() => Table().distributedBy_ != null;
+        public bool IsDistributionMatch(List<Expr> keys, QueryOption option)
+        {
+            // check if the list of expressions match
+            // the distribution of this table
+            if (!IsDistributed())
```
Comment (zhouqingqing) [inline qpmodel/DataType.cs:268]: please clarify this comment: only distributed deployment shall check it

File: qpmodel/DataType.cs [hunk 2]
```diff
@@ -261,6 +261,23 @@ public BaseTableRef(string name, string alias = null, SelectStmt.TableSample tab
...
+                return true;
+            else
+            {
+                // check if number of partitions agree
+                if (option.optimize_.query_dop_ == Table().distributions_.Count)
```
Comment (zhouqingqing) [inline qpmodel/DataType.cs:273]: let's assert dop = distribution count

File: qpmodel/LogicNode.cs [hunk 3]
```diff
@@ -585,27 +586,31 @@ void createOneKeyList(BinExpr fb)
                 Debug.Assert(leftKeys_.Count == rightKeys_.Count);
             }
 
-            // reset the old values
-            if (!canReUse)
+            // make sure multithreading does not mess up shared logic node
```
Comment (zhouqingqing) [inline qpmodel/LogicNode.cs:589]: can you clarify this?

File: qpmodel/LogicNode.cs [hunk 4]
```diff
@@ -585,27 +586,31 @@ void createOneKeyList(BinExpr fb)
...
+            lock (this)
```
Comment (zhouqingqing) [inline qpmodel/LogicNode.cs:590]: adjust order to avoid this?

File: qpmodel/PhysicalNode.cs [hunk 5]
```diff
@@ -1431,12 +1431,20 @@ public class PhysicInsert : PhysicNode
 
         public override string Exec(Func<Row, string> callback)
         {
+            var table = (logic_ as LogicInsert).targetref_.Table();
+            var dop = context_.option_.optimize_.query_dop_;
+            if (table.distributedBy_ != null)
```
Comment (arzuschen) [inline qpmodel/PhysicalNode.cs:1436]: put assertion in create table ref

**Pull request merged**